tests/compile-test.rs

```diff
@@ -46,50 +56,38 @@ fn third_party_crates() -> String {
-    let mut flags = String::new();
-    for entry in fs::read_dir(dep_dir).unwrap().flatten() {
-        let path = entry.path();
-        if let Some(name) = try {
-            let name = path.file_name()?.to_str()?;
-            let (name, _) = name.strip_suffix(".rlib")?.strip_prefix("lib")?.split_once('-')?;
-            CRATES.iter().copied().find(|&c| c == name)?
-        } {
-            flags += &format!(" --extern {}={}", name, path.display());
-            crates.entry(name).or_default().push(path.clone());
-        }
-    }
-    crates.retain(|_, paths| paths.len() > 1);
-    if !crates.is_empty() {
-        let crate_names = crates.keys().map(|s| format!("`{}`", s)).collect::<Vec<_>>().join(", ");
-        // add backslashes for an easy copy-paste `rm` command
-        let paths = crates
-            .into_values()
-            .flatten()
-            .map(|p| strip_current_dir(&p).display().to_string())
-            .collect::<Vec<_>>()
-            .join(" \\\n");
-        // Check which action should be done in order to remove compiled deps.
-        // If pre-installed version of compiler is used, `cargo clean` will do.
-        // Otherwise (for bootstrapped compiler), the dependencies directory
-        // must be removed manually.
-        let suggested_action = if std::env::var_os("RUSTC_BOOTSTRAP").is_some() {
-            "removing the stageN-tools directory"
-        } else {
-            "running `cargo clean`"
-        };
 
-        panic!(
-            "\n----------------------------------------------------------------------\n\
-            ERROR: Found multiple rlibs for crates: {}\n\
-            Try {} or remove the following files:\n\n{}\n\n\
-            For details on this error see https://github.com/rust-lang/rust-clippy/issues/7343\n\
-            ----------------------------------------------------------------------\n",
-            crate_names, suggested_action, paths
-        );
+    let depinfo = {
+        let exe_path = env::current_exe().unwrap();
+        let depinfo_filename = format!("{}.d", exe_path.file_stem().unwrap().to_str().unwrap());
+        let mut dep_info_path = exe_path;
+        dep_info_path.set_file_name(depinfo_filename);
+        std::fs::read_to_string(dep_info_path).unwrap()
+    };
+    let mut crates: HashMap<&str, &str> = HashMap::with_capacity(CRATES.len());
+    for line in depinfo.lines() {
+        let path = try {
+            let path = line.strip_suffix(':')?;
+            let part = path
+                .strip_suffix(".rlib")
+                .or_else(|| path.strip_suffix(".so"))
+                .or_else(|| path.strip_suffix(".dylib"))
+                .or_else(|| path.strip_suffix(".dll"))?;
+            let part = part.rsplit_once('-')?.0;
+            let name = part.rsplit_once(|c| matches!(c, '/' | '\\'))?.1.strip_prefix("lib")?;
+            CRATES.contains(&name).then(|| (name, path))?
```

Comment:
This seems to make really strong assumptions about the format of the depinfo file. Since this is still unstable this may just change without a warning. Also reading the tracking issue https://github.com/rust-lang/rust/issues/63012, I wonder if this also works on Windows?

Reply:
Well let's spell out those assumptions (I should put this in a comment too). The depinfo file is a Makefile. For every rlib, there is a "rule" where the "target" is the rlib file and there are no "prerequisites" after the `:`. There is no extra whitespace around the `:`. On the plus side there is no assumption about where the path is in the filesystem.

I agree these are pretty strong assumptions. But I also think this buys us much better stability overall than the current solution. The file is machine-generated and seems to be unchanged for 2 years, which is quite good in this ecosystem. The flag is also used in rustc builds so that's a plus.

> I wonder if this also works on Windows?

I *think* it will since I allow for either slash and don't look at the beginning of the path. But someone should test this before merging for sure.

Reply:
> (I should put this in a comment too)

Yep :D

>  The file is machine-generated and seems to be unchanged for 2 years

Sounds good!

> I _think_ it will since I allow for either slash and don't look at the beginning of the path. But someone should test this before merging for sure.

@rust-lang/clippy anyone has access/uses a windows PC where they can test this?

Reply:
You'll need to do:

```rust
let part = part.rsplit_once('-')?.0.rsplit_once(|c| matches!(c, '/' | '\\'))?.1;
let name = part.strip_prefix("lib").unwrap_or(part);
```

dll names on windows don't have a lib prefix.